executor/mpp_gather.go

```diff
@@ -0,0 +1,331 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/mpp"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/plancodec"
+	"github.com/pingcap/tidb/util/ranger"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// Currently we only use mpp for broadcast join.
+func useMPPExecution(ctx sessionctx.Context, tr *plannercore.PhysicalTableReader) bool {
+	if !ctx.GetSessionVars().AllowMPPExecution {
+		return false
+	}
+	if tr.StoreType != kv.TiFlash {
+		return false
+	}
+	return true
```

Comment:
Should we return `true` only when there is a `PhysicalBCJ` in `tr.TablePlans`?

Reply:
I plan to use optimizer + flag to check if we use mpp.
Currenty let's go around it